### DISCUSSION TITLE
ios 15 Cannot install adhoc build app.

### DISCUSSION THREAD

**Discussion by avelle** [airsdk/Adobe-Runtime-Support#1063] [Q&A]
''xxxx needs to be updated'' prompt is noticed upon installing adhoc build.

**Comment by SUSDEV** [comment]
Harman has not released the SDK that will work with iOS 15, but it [sounds like they are working on it and it is coming out hopefully soon](https://github.com/airsdk/Adobe-Runtime-Support/discussions/899).

It is so frustrating that Apple has not given warning to developers about what is breaking, and Apple is releasing iOS 15 in maybe a month or so and the next release of Xcode is not final yet, to give Harman assurance. At least when Apple cleared the app store of 32-bit apps they gave a lot of notice. I have started preparing my customers not to update to iOS 15 until I have the app available (at this point, even 'if' as we don't know what Apple has broken from 14 to 15). Of course thanks to Apple I don't really know all of my customers except for the ones who have opted in. I guess a bunch of early Christmas presents from Apple...not!

**Reply by marchbold** [reply]
Apple normally handles this transition quite well, with apps that are packaged against an sdk run in a "compatibility mode". 

So I wouldn't stress too much about when iOS 15 is released. The important date is when Apple starts requiring apps to be built with iOS 15, which normally occurs several months after the OS release.

**Reply by ajwfrost** [reply]
Hi @avelle et al

From what we can see, there's actually a code signing issue with our current IPA files, and we're not sure whether Apple will allow the existing IPAs to be installed onto iOS15. It looks like Xcode 12.5 signing already includes the entitlements both as an embedded plist (which we do) and also as a DER-encoded blob (which we didn't do).

When we're resigning apps, we can only get them to install onto iOS15 if we include the `--generate-entitlement-der` option in the `codesign` parameters; @marchbold you may want to update your re-signing script to include this if you hadn't already!

Naturally, only after a lot of research by one of our developers have we managed to work this out, and once we found the solution we then found a source that confirmed it (which would have been good to find a little earlier...!): https://stackoverflow.com/questions/68467306/the-code-signature-version-is-no-longer-supported

Anyway: we have updated ADT now so that it uses this additional parameter (if code-signing using `codesign` e.g. on a macOS with recent JRE) and the internal code-signing implementation is generating the DER blob and additional hashes..

It's working with our simple test cases, but I would be grateful if anyone can share their experiences! You can update ADT with the below link to overwrite the 33.1.1.575 SDK version.

Note, we will also have a build of the AIR SDK that includes all the new frameworks/libraries and Xcode 13 fields, as soon as the 'final' Xcode 13 and SDK for iphoneos 15 is released...

thanks

https://transfer.harman.com/message/Hn0uKXIqQQHARz1dRhERvH

**Reply by marchbold** [reply]
@ajwfrost We don't use that script anymore since you got dynamic framework signing working, but I'll get it updated at some point.

**Reply by SUSDEV** [reply]
Hi @ajwfrost does "the internal code-signing" mean we should not have to codesign the IPA after we generate it with our development tool (FDT, in my case)? I successfully deployed my app to TestFlight using 33.1.1.575 for iOS 14.7.1 without having to do some extra codesigning.

I have not updated my iPad pro to 15 beta yet but it sounds like it's time. I guess I will try an ad hoc distribution though it seems that goes through less stringent filtering than a production for app store build. I will let you know when I try it hopefully in the next few days.

**Reply by ajwfrost** [reply]
@SUSDEV the "internal" codesigning is code that's built into the ADT tool to directly generate a code signature - i.e. what Adobe had created initially, and what we've had to enhance to support the latest code-signing requirements, which has been one of the more complex pieces of work to try to keep up with their undocumented requirements! So ADT can also now support a mode where it uses the native "codesign" utility that's available on a Mac, which means it /should/ be more future-proof (although, we did just find that it needs another parameter as of Xcode 14.5...).

So either way: you should not have to use an external script or to re-codesign the IPA file or similar. If you find that you do need to do it -> then we'd like to have the "before" and "after" versions of the app so that we can see where our code is missing something...

I'm a little surprised that an existing AIR-generated IPA file is working on a Beta 15 device, we couldn't get that to work, until we'd made these recent changes for it!

**Reply by SUSDEV** [reply]
Hi @ajwfrost I would love to take the 'win' as is but it makes me really nervous as to the real status once iOS 15 rolls out. Do you think it would be worthwhile for me to email Apple Technical Development help or they won't really know either? Is there something I can send you if you want to inspect the ipa? I hate leaving this in limbo.

**Reply by ajwfrost** [reply]
Hi @SUSDEV - to be honest, I don't fully know, but I would hope that for App Store deployments, it would be okay. The change in code signature was introduced in Xcode 12.5, so if Apple are being strict about this then anything that people created with Xcode 12.4 would fail to load on iOS 15.. not sure whether they would be that strict though!

In terms of your actual IPA file: did you do any re-signing of this to get it working before you uploaded it? and if so, what machine did you use for this (i.e. the macOS version of the machine that was running codesign)?

If you're able to run the "codesign -dvvvv MyApp.app" command on your app, we can give more details (unzip the .ipa file and then you'll see the .app folder). Or just send it to me :-)

thanks

**Reply by SUSDEV** [reply]
@ajwfrost I think 33.1.1.216 used 12.1 so this just might be my lucky day with Apple to buy me more time.

I didn't do any extra codesigning on the ipa.

**Reply by kalashot** [reply]
Hi, Andrew 
Could you please share a link to adt.jar once again? This link ( https://transfer.harman.com/message/Hn0uKXIqQQHARz1dRhERvH) doesn't work. Thank you.

**Reply by ajwfrost** [reply]
Yes - I wish our system allowed links to be valid for more than two weeks! Please use the below...
https://transfer.harman.com/message/tDd168zzA6P4iVqBbn1iyh
thanks

**Reply by kalashot** [reply]
Thanks for the quick response! Our enterprise app installs/runs well on ios 15 now! Thank you so much again :)

**Comment by SUSDEV** [comment]
I just upgraded my iPad Pro 2019 to iOS 15 Beta 7. My app (made with 33.1.1.216) is working fine as-is, downloaded from the App Store. I haven't tested all the functions. Has anyone else tried Beta 7? Of course at some point I want to update the app, but the urgency is less at this point.

**Reply by SUSDEV** [reply]
@ajwfrost Would someone be willing to help me verify this older version works on a beta 15 device? The app is called SimsUshare. It cost $9.99  I would reimburse someone who would try it just msg me please. This is super important!

**Comment by GingerAle0426** [comment]
Let me show the report of our environment at 27th Aug 2021.

I have tried to install "Store app" and "In house(ad hoc) App"  on my iPad with iPad OS15β.
They are both build with AIR SDK 33.1.1.575.

We have successful installation of "Store app",but doesn't about "In house App" with the prompt "'xxxx needs to be updated".

This problem haven't been resolved about "In house App".

Is there anything for us to find the cause?

**Reply by ajwfrost** [reply]
Hi @GingerAle0426 thanks for this, that's pretty interesting to see the difference. Were you using the normal 33.1.1.575 version, or did you patch the adt.jar file?
If you didn't - then please update it from the below link (perhaps after backing up the existing one) and that should help with the ad-hoc signing..
https://transfer.harman.com/message/Hn0uKXIqQQHARz1dRhERvH
thanks
  Andrew

p.s. @SUSDEV this is another indication that Apple are a little more forgiving for applications that are installed from the App Store vs ones that are deployed more manually...

**Reply by GingerAle0426** [reply]
Thank you @ajwfrost!

By overwriting the adt.jar on latest AIR SDK ,I succeeded in installing the app on the iPadOS 15.
This have let my clients (using enterprise apps) free from the fear of iPadOS updating,thank you.

I also got the same result with the method of Code Resigning on Mac, which I learned in another forum about Apple's changing policy in 2020.
I found out that they have the same effect.

We look forward to the patch-free AIR SDK after the official release of the iPadOS 15.

**Reply by ajwfrost** [reply]
Great, thanks for confirming that @GingerAle0426 

Yes, code resigning would work as well, either with the `--generate-entitlement-der` option, or potentially that's enabled by default on recent macOS versions...

**Comment by PhilCase** [comment]
Hi is the patch file still available please?    I have this issue currently.    The hyperlink has expired.

**Reply by ajwfrost** [reply]
The patch went into the .620 release which is just set up as a beta currently, there's another release going through the pipeline at the moment that adds in the final changes needed for iOS15.

But for that ADT file then you can get it from the .620 release:
https://airsdk.harman.com/download/33.1.1.620
(you can use this for production, we didn't trigger it as the main/latest release just because the forthcoming one will supersede it..)

thanks

**Reply by PhilCase** [reply]
Thanks Andrew. 

I'd tried against the .620 release.   I was trying to make an adhoc installer.    It would build in .575 but would not build in .620.    
I'll try again.

**Reply by PhilCase** [reply]
I've made .620 libraries in the same way I made the other libraries.    I take the "previous working set" and copy the .620 libraries over the top.     

IntelliJ Idea doesn't allow me to include that set of libraries in my project and requires that I go back to .575.   It actually doesn't allow me to save the change of SDK for the project.

**Reply by ajwfrost** [reply]
Hi @PhilCase  just coming back to this one .. I'm not entirely clear on your last comment above, have you been able to get a build working with IntelliJ and the 33.1.1.620 AIR SDK? I wasn't sure if you were using the Flex one which has its own complexities, but it should be possible to copy the normal SDKs over the top of each other (always assuming there's no other application open that's holding a lock on one of the files...)

thanks

**Reply by PhilCase** [reply]
I've taken a copy of my working .575 libraries and have copied the patch for 620 over the top of it.    (AIRSDK_Flex_Windows.zip)
I can create Android Apps with it.

If I build for iOS against the 575 libraries it builds but does not deploy.   I get the same issue with the DER that everyone else has reported.

If I swap out the SDK for my .620 SDK it doesn't build.    I get the following error:

![image](https://user-images.githubusercontent.com/11427050/135439026-b1aed026-b6f7-4110-9c25-c2a7e22ef336.png)

against a "limited distrobution" build for internal testing.

.620 seems to build Android apps.

I know I can make release builds for TestFlight.     

It will be best for me to wait for the full release library with the iOS fix as I have other issues I'm trying to fix.

I also cannot use debug builds for iOS currently.    I'm not sure if these are "connection issues" as "adt -devices -platform ios" doesn't always seem to connect to the device.   It's more likely to if I have iTunes running, but not guaranteed.

**Reply by ajwfrost** [reply]
Ah okay thanks -> that LinkedList/PlistNode issue was a bug in 620 that is fixed in the forthcoming 633 release.

Regarding device connection etc, we're using a third party library for this now so you can always go to the source of that and see whether it's working better there (which shouldn't be the case, but there are sometimes some weird permission things that can cause problems). So in a command prompt window (assuming a Windows machine per your comment above):
```
cd [AIRSDK_folder]\lib\aot\bin\iOSBin
ideviceinfo.exe
ideviceinstaller -i application.ipa
```
The `ideviceinfo` tells us about connected devices...

If you check back tomorrow, the next release of the AIR SDK should be available by then.

**Comment by Tron307020** [comment]
I am experiencing the same (or similar) problem with Ad-Hoc distribution on iOS 15. Upon an attempted install (ad-hoc), I am receiving the error: "[APP] Needs To Be Updated" (bold title) and "The developer of this app needs to update it to work with this version of iPadOS." Previous iOS versions work fine. I am using the latest SDK 33.1.1.633.

I had thought that updating the adt.jar file would work as it did for others here, but (1) none of the links to download the patch work anymore and (2) it was noted that the updated adt.jar file would be in release 620 and so also 633.

Any thoughts?
